arrow-fx-coroutines/src/main/kotlin/arrow/fx/coroutines/Enviroment.kt

```diff
@@ -0,0 +1,86 @@
+package arrow.fx.coroutines
+
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.CoroutineContext
+import kotlin.coroutines.startCoroutine
+
+/**
+ * Environment that can run [suspend] programs using [startCoroutine] and [startCoroutineCancellable].
+ *
+ * An [Environment] runs on a certain [CoroutineContext] which is used to start the programs on.
+ * Since coroutines always return where they were started, this [CoroutineContext] also defines where you return after operators like [sleep] or [evalOn].
+ * Therefore it's advised to always run on [ComputationPool] which is the default setting.
```

Comment:
Shouldn't we default to `EmptyCoroutineContext` as the kx coroutines builders like `async` or `launch`? Thinking on Android for example, you always wanna start stuff on the main thread then you decide which effects you want to offload to specific contexts like `ComputationPool` or `IOPool` with `evalOn` per occurrence, but you always want to get back to the caller thread which in Android that's the Main one. (For the majority of the cases). Also outside of Android, doesn't it make more sense to default to a no-op context like `EmptyCoroutineContext` so you stay on the caller thread for predictability?

Reply:
No, we cannot default to `EmptyCoroutineContext` since that's an empty map and thus doesn't have a `ContinuationInterceptor`. As a result when you call an operator such as `evalOn` there is no way to intercept to return to the original context. In order for that to be possible we'd need launch an `EventLoop` on the current context, so we can later return there. Which of course also comes with a cost.

I.e. 

```
//main w/ EmptyCoroutineContext
evalOn(ComputationPool) {
  // forkjoinpool
} <-- // Can't go back to main since `EmptyCoroutineContext` cannot dispatch us back there
```

> you always wanna start stuff on the main thread

Typically on Android View setup needs to happen as fast as possible, and eagerly. Often this also has to happens in a different callbacks than where you call your business logic. I.e. `onCreateView`, `onViewCreated` & `onResume`.
So instead I prefer to launch my business logic on `ComputationPool`, and return to the `UI` if/when I need to using `evalOn(UI)`.

Alternatively if you can also launch on `Main` CoroutineContext for Android which would come at the price of a `post`. Unless you implement it with a check to see if it's already on `Main` then it would only `post` when returning from a different context like in the above example. We can however not default to `Android` Main here, but we could do so in an `AndroidEnviroment` in an Android module. Where can also provide nicer APIs for Android.

```
interface AndroidEnviroment : Enviroment {
  fun unsafeRunScoped(lifecycleOwner: LifecycleOwner, ...)
}
```

Reply:
Got it. Regardless Android, I'm used to stay on caller thread until there's something computationally or IO expensive that requires jumping off to other pools in my program. Same for tests, you start execution on the test thread and then at some point and only if needed, you can jump off for a given effect (evalOn). Here, we do the other way around. We assume that anybody that runs a program by default (and uses `Environment`) wants the program to run in a `ComputationPool` by default and just jump off of this pool explicitly on some arbitrary places. I'm not 100% sure about that since feel like assuming a thread jump at the beginning for the user. But I believe we've had this conversation many times, so not a blocker for this PR at all 👍

Reply:
This pattern is a pattern commonly on back-end. Maybe @raulraja can shed some more light.

We should also stay warry not to let any front-end concerns leak into Arrow Fx Coroutines. Therefore I intentionally left the implementation open for composition through an interface, the default implementation is a simple wrapper of the primitive runners.

Running your program on `Main` or similar single thread contexts comes with a hidden drawback that it's easy to keep it occupied with long computations without realizing it until you see freezing in the app, no?

Reply:
> Running your program on `Main` or similar single thread contexts comes with a hidden drawback that it's easy to keep it occupied with long computations without realizing it until you see freezing in the app, no?

In general yes, on Android specifically there are logs and checks to prevent that by either crashing the app or by showing some cryptic logs... 

I agree with @nomisRev on this because by defaulting to a computation thread we eliminate the burden of the dev to remember to switch threads, because not doing so might or **might not** be noticeable during testing, but might come up on real and slower devices. So you only need to "remember" to post on UI, which you can very easily automate by creating some `evalOn(UI) { post(state) }` kind of functions.

Reply:
Sounds good, thanks for sharing thoughts folks 👍